In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [ ]:
file_path = '../../data/Prac1/breast-cancer-wisconsin.data'
columns = ['sample_code', 'clump_thickness', 'cell_size', 'cell_shape', 'marginal_adhesion', 'single_cell_size', 'bare_nuclei', 'bland_chromatin', 'normal_nucleoli', 'mitoses', 'class']
df = pd.read_csv(file_path, header=None, names=columns)
df

sample_code  clump_thickness  cell_size  cell_shape  marginal_adhesion  \
0        1000025                5          1           1                  1   
1        1002945                5          4           4                  5   
2        1015425                3          1           1                  1   
3        1016277                6          8           8                  1   
4        1017023                4          1           1                  3   
..           ...              ...        ...         ...                ...   
694       776715                3          1           1                  1   
695       841769                2          1           1                  1   
696       888820                5         10          10                  3   
697       897471                4          8           6                  4   
698       897471                4          8           8                  5   

     single_cell_size bare_nuclei  bland_chromatin  normal_nucleoli  mitoses  \
0                   2           1                3                1        1   
1                   7          10                3                2        1   
2                   2           2                3                1        1   
3                   3           4                3                7        1   
4                   2           1                3                1        1   
..                ...         ...              ...              ...      ...   
694                 3           2                1                1        1   
695                 2           1                1                1        1   
696                 7           3                8               10        2   
697                 3           4               10                6        1   
698                 4           5               10                4        1   

     class  
0        2  
1        2  
2        2  
3        2  
4        2  
..     ...  
694      2  
695      2  
696      4  
697      4  
698      4  

[699 rows x 11 columns]

In [3]:
df.describe()

sample_code  clump_thickness   cell_size  cell_shape  \
count  6.990000e+02       699.000000  699.000000  699.000000   
mean   1.071704e+06         4.417740    3.134478    3.207439   
std    6.170957e+05         2.815741    3.051459    2.971913   
min    6.163400e+04         1.000000    1.000000    1.000000   
25%    8.706885e+05         2.000000    1.000000    1.000000   
50%    1.171710e+06         4.000000    1.000000    1.000000   
75%    1.238298e+06         6.000000    5.000000    5.000000   
max    1.345435e+07        10.000000   10.000000   10.000000   

       marginal_adhesion  single_cell_size  bland_chromatin  normal_nucleoli  \
count         699.000000        699.000000       699.000000       699.000000   
mean            2.806867          3.216023         3.437768         2.866953   
std             2.855379          2.214300         2.438364         3.053634   
min             1.000000          1.000000         1.000000         1.000000   
25%             1.000000          2.000000         2.000000         1.000000   
50%             1.000000          2.000000         3.000000         1.000000   
75%             4.000000          4.000000         5.000000         4.000000   
max            10.000000         10.000000        10.000000        10.000000   

          mitoses       class  
count  699.000000  699.000000  
mean     1.589413    2.689557  
std      1.715078    0.951273  
min      1.000000    2.000000  
25%      1.000000    2.000000  
50%      1.000000    2.000000  
75%      1.000000    4.000000  
max     10.000000    4.000000

In [4]:
df.replace('?', np.nan, inplace=True)

In [5]:
df.isna().sum()

sample_code           0
clump_thickness       0
cell_size             0
cell_shape            0
marginal_adhesion     0
single_cell_size      0
bare_nuclei          16
bland_chromatin       0
normal_nucleoli       0
mitoses               0
class                 0
dtype: int64

In [6]:
df['bare_nuclei'] = pd.to_numeric(df['bare_nuclei'], errors='coerce')

In [7]:
df['bare_nuclei'].fillna(df['bare_nuclei'].median(), inplace=True)

C:\Users\MY PC\AppData\Local\Temp\ipykernel_22692\290591324.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['bare_nuclei'].fillna(df['bare_nuclei'].median(), inplace=True)


In [10]:
df_2 = df[df['class'] == 2]
df_2.shape

(458, 11)

In [11]:
df_4 = df[df['class'] == 4]
df_4.shape

(241, 11)

In [72]:
random_state = 18

test_set = pd.concat([df_2.sample(n=80, random_state=random_state), df_4.sample(n=40, random_state=random_state)])
test_set 

sample_code  clump_thickness  cell_size  cell_shape  marginal_adhesion  \
136      1182404                4          1           1                  1   
597      1333063                5          1           3                  1   
516      1320077                1          1           1                  1   
407      1234554                1          1           1                  1   
411      1238464                1          1           1                  1   
..           ...              ...        ...         ...                ...   
56       1112209                8         10          10                  1   
39       1091262                2          5           3                  3   
308       667204                7          8           7                  6   
668      1350319                5          7           4                  1   
254      1116116                9         10          10                  1   

     single_cell_size  bare_nuclei  bland_chromatin  normal_nucleoli  mitoses  \
136                 2          1.0                2                1        1   
597                 2          1.0                3                1        1   
516                 1          1.0                1                1        1   
407                 2          1.0                2                1        1   
411                 1          1.0                2                1        1   
..                ...          ...              ...              ...      ...   
56                  3          6.0                3                9        1   
39                  6          7.0                7                5        1   
308                 4          3.0                8                8        4   
668                 6          1.0                7               10        3   
254                10          8.0                3                3        1   

     class  
136      2  
597      2  
516      2  
407      2  
411      2  
..     ...  
56       4  
39       4  
308      4  
668      4  
254      4  

[120 rows x 11 columns]

In [73]:
train_set = pd.concat([df_2.drop(df_2.sample(n=80, random_state=random_state).index), df_4.drop(df_4.sample(n=40, random_state=random_state).index)])
train_set

sample_code  clump_thickness  cell_size  cell_shape  marginal_adhesion  \
2        1015425                3          1           1                  1   
3        1016277                6          8           8                  1   
4        1017023                4          1           1                  3   
6        1018099                1          1           1                  1   
8        1033078                2          1           1                  1   
..           ...              ...        ...         ...                ...   
681      1371026                5         10          10                 10   
691       695091                5         10          10                  5   
696       888820                5         10          10                  3   
697       897471                4          8           6                  4   
698       897471                4          8           8                  5   

     single_cell_size  bare_nuclei  bland_chromatin  normal_nucleoli  mitoses  \
2                   2          2.0                3                1        1   
3                   3          4.0                3                7        1   
4                   2          1.0                3                1        1   
6                   2         10.0                3                1        1   
8                   2          1.0                1                1        5   
..                ...          ...              ...              ...      ...   
681                 4         10.0                5                6        3   
691                 4          5.0                4                4        1   
696                 7          3.0                8               10        2   
697                 3          4.0               10                6        1   
698                 4          5.0               10                4        1   

     class  
2        2  
3        2  
4        2  
6        2  
8        2  
..     ...  
681      4  
691      4  
696      4  
697      4  
698      4  

[579 rows x 11 columns]

In [74]:
for column in columns[1:]:
    print(train_set[column].unique())

[ 3  6  4  1  2  5  8  7 10  9]
[ 1  8  2  6  3  4  9  7 10  5]
[ 1  8  3  2  6  4  5  7 10  9]
[ 1  3  2  9  5  4  6 10  8  7]
[ 2  3  1  6  4 10  8  5  7  9]
[ 2.  4.  1. 10.  3.  5.  7.  8.  9.  6.]
[ 3  1  2  7  4  5  6  8 10  9]
[ 1  7  8  2  3  6  5  9  4 10]
[ 1  5  2  7  3  8  4 10  6]
[2 4]


In [75]:
X_train = train_set.drop('class', axis='columns')
y_train = train_set['class']

In [76]:
X_test = test_set.drop('class', axis='columns')
y_test = test_set['class']

In [77]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 579 entries, 2 to 698
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sample_code        579 non-null    int64  
 1   clump_thickness    579 non-null    int64  
 2   cell_size          579 non-null    int64  
 3   cell_shape         579 non-null    int64  
 4   marginal_adhesion  579 non-null    int64  
 5   single_cell_size   579 non-null    int64  
 6   bare_nuclei        579 non-null    float64
 7   bland_chromatin    579 non-null    int64  
 8   normal_nucleoli    579 non-null    int64  
 9   mitoses            579 non-null    int64  
dtypes: float64(1), int64(9)
memory usage: 49.8 KB


In [78]:
from sklearn.naive_bayes import GaussianNB

model = GaussianNB()
model.fit(X_train, y_train)

GaussianNB()

In [79]:
model.score(X_test, y_test)

0.8916666666666667

In [80]:
y_pred = model.predict(X_test)
y_pred

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 4, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 4, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 4, 2, 2, 2, 2, 2,
       2, 2, 2, 4, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 4, 2, 2, 4, 4, 4, 2, 4,
       4, 4, 4, 2, 4, 4, 4, 4, 4, 4, 4, 4, 4, 2, 2, 2, 4, 4, 4, 4, 4, 4,
       4, 4, 2, 4, 4, 4, 4, 4, 2, 4])

In [81]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score

In [82]:
accuracy_score(y_test, y_pred)

0.8916666666666667

In [83]:
precision_score(y_test, y_pred, average='macro')

0.8899159663865546

In [84]:
recall_score(y_test, y_pred, average='macro')

0.8625

In [85]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           2       0.89      0.95      0.92        80
           4       0.89      0.78      0.83        40

    accuracy                           0.89       120
   macro avg       0.89      0.86      0.87       120
weighted avg       0.89      0.89      0.89       120



**Lớp 2:**

* Precision = 0.89 → Trong số các mẫu dự đoán là "2", có 89% thực sự là "2".
* Recall = 0.95 → 95% số mẫu thực sự là "2" đã được mô hình phát hiện đúng.
* F1-score = 0.92 → Hiệu suất tổng thể tốt.

**Lớp 4:**

* Precision = 0.89 → Trong số các mẫu dự đoán là "4", có 89% thực sự là "4".
* Recall = 0.78 → 78% số mẫu thực sự là "4" đã được mô hình phát hiện đúng.
* F1-score = 0.83 → Tổng thể khá.

### Các chỉ số tổng hợp:

* Accuracy (Độ chính xác tổng thể) = 0.89 (~89%)
    * Tỷ lệ dự đoán đúng trên tổng số mẫu.​
 
* Macro avg (Trung bình không cân bằng) = 0.89, 0.86, 0.87
    * Trung bình cộng của Precision, Recall và F1-score của từng lớp.
* Weighted avg (Trung bình có cân nhắc số lượng mẫu) = 0.89, 0.89, 0.89
    * Giống macro avg nhưng có tính đến số lượng mẫu (support).